File: pipeline/analyze/tw_patch_skus.py
#!/usr/bin/env python3
"""
Patches pages 35-43 in totalwine_lexington_green.jsonl with
SKU, size, and URL data from the browser localStorage export.

Usage:
    python -m pipeline.analyze.tw_patch_skus
"""

import json
import re
import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parents[2]))

JSONL_FILE = Path(__file__).resolve().parents[2] / "totalwine_lexington_green.jsonl"
PATCH_FILE = Path("C:/Users/neilw/Downloads/tw_full_data.txt")


def parse_categories_from_url(url: str) -> list[str]:
    categories = []
    if not url:
        return categories
    cleaned = re.sub(r"\?.*$", "", url).replace("/wine/", "", 1)
    parts = cleaned.strip("/").split("/")
    skip = {"deals", "gift-center", "p"}
    for i, part in enumerate(parts):
        if part in skip or re.match(r"^\d+$", part):
            continue
        if i == len(parts) - 1 or (i == len(parts) - 3 and parts[-2] == "p"):
            continue
        readable = " ".join(w.capitalize() for w in part.split("-"))
        categories.append(readable)
    return categories


def main():
    # Build lookup from patch file
    patch_map: dict[str, dict] = {}
    if PATCH_FILE.exists():
        for line in PATCH_FILE.read_text(encoding="utf-8").strip().split("\n"):
            parts = line.split("|")
            if len(parts) >= 4 and parts[0]:
                patch_map[parts[0].strip()] = {
                    "sku": parts[1] or "", "size": parts[2] or "", "url": parts[3] or "",
                }
    else:
        print(f"Patch file not found: {PATCH_FILE}")
        sys.exit(1)

    print(f"Loaded {len(patch_map)} patch entries")

    lines = JSONL_FILE.read_text(encoding="utf-8").strip().split("\n")
    patched = 0
    sku_added = 0
    size_added = 0
    url_added = 0

    updated = []
    for line in lines:
        obj = json.loads(line)
        if obj.get("page", 0) < 35:
            updated.append(line)
            continue

        patch = patch_map.get(obj.get("name"))
        if not patch:
            updated.append(line)
            continue

        patched += 1
        if patch["sku"] and not obj.get("sku"):
            obj["sku"] = patch["sku"]
            sku_added += 1
        elif patch["sku"] and obj.get("sku") != patch["sku"]:
            obj["sku"] = patch["sku"]
        if patch["size"] and not obj.get("size"):
            obj["size"] = patch["size"]
            size_added += 1
        if patch["url"] and not obj.get("url"):
            obj["url"] = patch["url"]
            url_added += 1
        if patch["url"] and not obj.get("categories"):
            obj["categories"] = parse_categories_from_url(patch["url"])

        updated.append(json.dumps(obj))

    JSONL_FILE.write_text("\n".join(updated) + "\n", encoding="utf-8")
    print(f"Patched {patched} entries:")
    print(f"  SKUs added/updated: {sku_added}")
    print(f"  Sizes added: {size_added}")
    print(f"  URLs added: {url_added}")


if __name__ == "__main__":
    main()

File: scripts_archive/node/root/tw_patch_skus.mjs
#!/usr/bin/env node
/**
 * Patches pages 35-43 in totalwine_lexington_green.jsonl with
 * SKU, size, and URL data from the browser localStorage export.
 */
import { readFileSync, writeFileSync } from 'fs';

const JSONL_FILE = 'totalwine_lexington_green.jsonl';
const PATCH_FILE = 'C:/Users/neilw/Downloads/tw_full_data.txt';

// Build lookup from patch file: name -> {sku, size, url}
const patchLines = readFileSync(PATCH_FILE, 'utf8').trim().split('\n');
const patchMap = new Map();
for (const line of patchLines) {
  const [name, sku, size, url, page] = line.split('|');
  if (name) {
    patchMap.set(name.trim(), { sku: sku || '', size: size || '', url: url || '' });
  }
}
console.log(`Loaded ${patchMap.size} patch entries`);

// Parse category from URL path
function parseCategoriesFromUrl(url) {
  const categories = [];
  if (!url) return categories;
  const cleaned = url.replace(/\?.*$/, '').replace(/^\/wine\//, '');
  const parts = cleaned.split('/');
  const skip = new Set(['deals', 'gift-center', 'p']);
  for (const part of parts) {
    if (skip.has(part)) continue;
    if (/^\d+$/.test(part)) continue;
    const idx = parts.indexOf(part);
    if (idx === parts.length - 1 || (idx === parts.length - 3 && parts[parts.length - 2] === 'p')) continue;
    const readable = part.split('-').map(w => w.charAt(0).toUpperCase() + w.slice(1)).join(' ');
    categories.push(readable);
  }
  return categories;
}

// Read and patch JSONL
const lines = readFileSync(JSONL_FILE, 'utf8').trim().split('\n');
let patched = 0, skuAdded = 0, sizeAdded = 0, urlAdded = 0;

const updated = lines.map(line => {
  const obj = JSON.parse(line);
  if (obj.page < 35) return line; // Leave pages 1-34 untouched

  const patch = patchMap.get(obj.name);
  if (!patch) return line;

  patched++;
  if (patch.sku && !obj.sku) { obj.sku = patch.sku; skuAdded++; }
  else if (patch.sku && obj.sku !== patch.sku) { obj.sku = patch.sku; } // update if different
  if (patch.size && !obj.size) { obj.size = patch.size; sizeAdded++; }
  if (patch.url && !obj.url) { obj.url = patch.url; urlAdded++; }
  if (patch.url && (!obj.categories || obj.categories.length === 0)) {
    obj.categories = parseCategoriesFromUrl(patch.url);
  }

  return JSON.stringify(obj);
});

writeFileSync(JSONL_FILE, updated.join('\n') + '\n');
console.log(`Patched ${patched} entries:`);
console.log(`  SKUs added/updated: ${skuAdded}`);
console.log(`  Sizes added: ${sizeAdded}`);
console.log(`  URLs added: ${urlAdded}`);
